File: src/DuneBot.Specs/Steps/StormSteps.cs
using DuneBot.Domain;
using DuneBot.Domain.State;
using DuneBot.Engine;
using DuneBot.Domain.Interfaces;
using Reqnroll;
using Moq;
using System.Collections.Generic;
using System.Linq;
using Xunit;
using System.Threading.Tasks;

namespace DuneBot.Specs.Steps
{
    [Binding]
    public class StormSteps
    {
        private readonly GameContext _context;

        public StormSteps(GameContext context)
        {
            _context = context;
        }

        private Game _game => _context.Game;
        private GameEngine _engine => _context.Engine;
        private DuneBot.Engine.Services.MapService _mapService => _context.MapService;

        [Given(@"a new game is starting")]
        public void GivenANewGameIsStarting()
        {
            _game.State.Turn = 1;
            _game.State.StormLocation = 18; // Standard start
        }

        [Given(@"the current storm position is sector (\d+)")]
        public void GivenTheCurrentStormPositionIsSector(int sector)
        {
            _game.State.StormLocation = sector;
        }
        
        [Given(@"the game is in Turn (\d+)")]
        public void GivenTheGameIsInTurn(int turn)
        {
            _game.State.Turn = turn;
        }

        [Given(@"the following forces are in ""(.*)"" \(Sector (\d+)\):")]
        public void GivenTheFollowingForcesAreIn(string territoryName, int sector, Table table)
        {
            var territory = _game.State.Map.Territories.FirstOrDefault(t => t.Name == territoryName);
            if (territory == null)
            {
                territory = new Territory { Name = territoryName, Sector = sector };
                _game.State.Map.Territories.Add(territory);
            }
            
            foreach (var row in table.Rows)
            {
                var factionName = row["Faction"];
                var count = int.Parse(row["Forces"]);
                var faction = (Faction)System.Enum.Parse(typeof(Faction), factionName);
                
                territory.FactionForces[faction] = count;
                
                if (!_game.State.Factions.Any(f => f.Faction == faction))
                {
                    _game.State.Factions.Add(new FactionState 
                    { 
                        Faction = faction, 
                        PlayerName = factionName,
                        PlayerDiscordId = (ulong)factionName.Length 
                    });
                }
            }
        }
        
        [Given(@"""(.*)"" \(Sector (\d+)\) has (\d+) Spice")]
        public void GivenTerritoryHasSpice(string territoryName, int sector, int amount)
        {
             var territory = _game.State.Map.Territories.FirstOrDefault(t => t.Name == territoryName);
             if (territory == null)
             {
                 territory = new Territory { Name = territoryName, Sector = sector };
                 _game.State.Map.Territories.Add(territory);
             }
             territory.SpiceBlowAmount = amount;
        }

        [When(@"the storm moves (\d+) sectors")]
        public void WhenTheStormMovesSectors(int amount)
        {
            // Bypass handler logic for specific move amount test
            int oldSector = _game.State.StormLocation;
            int newSector = _mapService.CalculateNextStormSector(oldSector, amount);
            _game.State.StormLocation = newSector;
            
            _context.BattleService.ApplyStormDamage(_game, oldSector, amount);
        }
        
        [When(@"the storm moves")]
        public async Task WhenTheStormMoves()
        {
            _game.State.Phase = GamePhase.Storm;
            
            var messageService = new DuneBot.Engine.Services.GameMessageService(); 
            var spiceService = new DuneBot.Engine.Services.SpiceService(_context.MockDeck.Object, messageService);
            
            var handler = new DuneBot.Engine.Phases.StormPhaseHandler(
                spiceService, 
                _context.BattleService, 
                _context.MapService);
                
            await handler.RunPhaseAsync(_game);
        }
        
        [When(@"the First Storm occurs")]
        public async Task WhenTheFirstStormOccurs()
        {
             await WhenTheStormMoves();
        }

        [When(@"the storm moves to sector (\d+)")]
        public void WhenTheStormMovesToSector(int sector)
        {
            _game.State.StormLocation = sector;
            
            // Trigger First Player Update manually using the handler
            var messageService = new DuneBot.Engine.Services.GameMessageService(); 
            var spiceService = new DuneBot.Engine.Services.SpiceService(_context.MockDeck.Object, messageService);
            var handler = new DuneBot.Engine.Phases.StormPhaseHandler(spiceService, _context.BattleService, _context.MapService);
            
            handler.UpdateFirstPlayer(_game);
        }
        
        [Given(@"the storm moves to sector (\d+)")]
        public void GivenTheStormMovesToSector(int sector)
        {
             _game.State.StormLocation = sector;
        }

        [Given(@"the players are seated as follows:")]
        public void GivenThePlayersAreSeatedAsFollows(Table table)
        {
             _game.State.Factions.Clear();
             
             int count = table.Rows.Count;
             int spacing = 18 / (count > 0 ? count : 1);
             
             // Assign Factions to sectors 1, 1+Spacing, etc.
             
             for (int i = 0; i < count; i++)
             {
                  var row = table.Rows[i];
                  var name = row["Faction"];
                  var faction = (Faction)System.Enum.Parse(typeof(Faction), name);
                  
                  // Assign StartSector
                  int sector = 1 + (i * spacing); 
                  
                  _game.State.Factions.Add(new FactionState
                  {
                      Faction = faction,
                      PlayerName = name,
                      PlayerDiscordId = (ulong)(i + 1),
                      StartSector = sector
                  });
             }
        }

        [Then(@"the new storm position should be sector (\d+)")]
        public void ThenTheNewStormPositionShouldBeSector(int expectedSector)
        {
            Assert.Equal(expectedSector, _game.State.StormLocation);
        }
        
        [Then(@"the new storm position should be between (\d+) and (\d+)")]
        public void ThenTheAnd(int min, int max)
        {
             Assert.InRange(_game.State.StormLocation, min, max);
        }

        [Then(@"""(.*)"" should have (\d+) forces in ""(.*)""")]
        public void ThenShouldHaveForcesIn(string factionName, int expectedCount, string territoryName)
        {
            var faction = (Faction)System.Enum.Parse(typeof(Faction), factionName);
            var territory = _game.State.Map.Territories.First(t => t.Name == territoryName);
            
            int actual = territory.FactionForces.ContainsKey(faction) ? territory.FactionForces[faction] : 0;
            Assert.Equal(expectedCount, actual);
        }
        
        [Then(@"""(.*)"" should represent (\d+) forces in the tanks")]
        public void ThenShouldRepresentForcesInTheTanks(string factionName, int expectedTanks)
        {
            var faction = (Faction)System.Enum.Parse(typeof(Faction), factionName);
            var fState = _game.State.Factions.First(f => f.Faction == faction);
            
            Assert.Equal(expectedTanks, fState.ForcesInTanks);
        }
        
        [Then(@"""(.*)"" should have (\d+) Spice")]
        public void ThenTerritoryShouldHaveSpice(string territoryName, int expectedSpice)
        {
             var territory = _game.State.Map.Territories.First(t => t.Name == territoryName);
             Assert.Equal(expectedSpice, territory.SpiceBlowAmount);
        }
        
        [Then(@"the First Player should be ""(.*)""")]
        public void ThenTheFirstPlayerShouldBe(string factionName)
        {
             var fpId = _game.State.FirstPlayerId;
             Assert.NotNull(fpId);
             
             var faction = _game.State.Factions.FirstOrDefault(f => f.PlayerDiscordId == fpId);
             Assert.NotNull(faction);
             Assert.Equal(factionName, faction.Faction.ToString());
        }
    }
}

File: ts-dune-bot/features/step_definitions/StormSteps.ts
import { Given, When, Then, Before } from '@cucumber/cucumber';
import { expect } from 'chai';
import { StormEngine } from '../../src/engine/StormEngine';
import { GameState, FactionState, TerritoryState } from '../../src/types';
import { prisma } from '../../src/db';

let lastError: any = null;
import { TestContext } from '../support/TestContext';

const stormEngine = new StormEngine();

const MOCK_TERRITORIES: any[] = [
    { name: "Old Gap", sector: 3, isSafe: false },
    { name: "Arrakeen", sector: 2, isSafe: true },
    { name: "Territory A", sector: 3, isSafe: false },
    { name: "Territory B", sector: 4, isSafe: false },
    { name: "Territory C", sector: 5, isSafe: false },
    { name: "Imperial Basin", sector: 6, isSafe: true },
    { name: "Sector 16 Sand", sector: 16, isSafe: false },
    { name: "Sector 17 Sand", sector: 17, isSafe: false },
    { name: "Sector 18 Sand", sector: 18, isSafe: false },
    { name: "Sector 1 Sand", sector: 1, isSafe: false },
    { name: "Sector 4 Sand", sector: 4, isSafe: false },
];

Before(() => {
    // No-op
});

// Helper to get state
async function getState(gameId: number): Promise<GameState> {
    const game = await prisma.game.findUnique({ where: { id: gameId } });
    if (!game) throw new Error("Game not found");
    return JSON.parse(game.stateJson) as GameState;
}

// Helper to save state
async function saveState(gameId: number, state: GameState) {
    await prisma.game.update({
        where: { id: gameId },
        data: { stateJson: JSON.stringify(state) }
    });
}

Given('a new game is starting', async function () {
    const initialState: GameState = {
        phase: "Setup",
        turn: 0,
        stormLocation: 0,
        factions: [],
        actionLog: [],
        auctionQueue: [],
        currentBid: 0,
        isBiddingRoundActive: false,
        spiceDeck: [],
        spiceDiscard: [],
        treacheryDeck: [],
        treacheryDiscard: [],
        nexusActive: false,
        boardState: {}
    };
    const game = await prisma.game.create({
        data: {
            guildId: "test-storm",
            stateJson: JSON.stringify(initialState),
            categoryId: "c", actionsChannelId: "a", mapChannelId: "m", tableTalkChannelId: "t"
        }
    });
    TestContext.gameId = game.id;
});

When('the First Storm occurs', async function () {
    const state = await getState(TestContext.gameId);
    // 0-20 sectors from Sector 18
    const initialMove = Math.floor(Math.random() * 21);
    state.stormLocation = ((18 + initialMove - 1) % 18) + 1;
    await saveState(TestContext.gameId, state);
});

Then('the new storm position should be between {int} and {int}', async function (min: number, max: number) {
    const state = await getState(TestContext.gameId);
    expect(state.stormLocation).to.be.gte(min);
    expect(state.stormLocation).to.be.lte(max);
});

Given('the current storm position is sector {int}', async function (location: number) {
    // If game doesn't exist, create it (reuse new game logic if needed, or assume sequential)
    if (TestContext.gameId === 0) {
        // Create dummy
        const initialState: GameState = {
            phase: "Setup",
            turn: 0, stormLocation: location, factions: [], actionLog: [], auctionQueue: [], currentBid: 0, isBiddingRoundActive: false, boardState: {},
            spiceDeck: [], spiceDiscard: [], treacheryDeck: [], treacheryDiscard: [], nexusActive: false
        };
        const game = await prisma.game.create({
            data: { guildId: "test-storm-2", stateJson: JSON.stringify(initialState), categoryId: "c", actionsChannelId: "a", mapChannelId: "m", tableTalkChannelId: "t" }
        });
        TestContext.gameId = game.id;
    } else {
        const state = await getState(TestContext.gameId);
        state.stormLocation = location;
        await saveState(TestContext.gameId, state);
    }
});



When('the storm moves', async function () {
    // Simulate random movement 1-3
    const state = await getState(TestContext.gameId);
    const move = Math.floor(Math.random() * 6) + 1;
    stormEngine.moveStorm(state, move, MOCK_TERRITORIES);
    state.stormMovedThisTurn = true;
    await saveState(TestContext.gameId, state);
});

When('the storm moves {int} sectors', async function (sectors: number) {
    const state = await getState(TestContext.gameId);
    stormEngine.moveStorm(state, sectors, MOCK_TERRITORIES);
    state.stormMovedThisTurn = true;
    await saveState(TestContext.gameId, state);
});

Given('the following forces are in {string} \\(Sector {int}):', async function (territoryName: string, sector: number, dataTable: any) {
    const state = await getState(TestContext.gameId);

    // Ensure BoardState territory exists
    if (!state.boardState[territoryName]) {
        state.boardState[territoryName] = { name: territoryName, spice: 0, forces: {} };
    }

    const rows = dataTable.hashes();
    for (const row of rows) {
        if (!state.boardState[territoryName].forces[sector]) {
            state.boardState[territoryName].forces[sector] = {};
        }
        state.boardState[territoryName].forces[sector][row['Faction']] = parseInt(row['Forces']);

        // Ensure faction exists in state for checks
        if (!state.factions.find(f => f.faction === row['Faction'])) {
            state.factions.push({
                faction: row['Faction'] as any,
                playerDiscordId: "d", playerName: row['Faction'],
                spice: 0, reserves: 0, forcesInTanks: 0, leaders: [], traitors: [], hand: []
            });
        }
    }
    await saveState(TestContext.gameId, state);
});

Given('{string} \\(Sector {int}) has {int} Spice', async function (territoryName: string, sector: number, amount: number) {
    const state = await getState(TestContext.gameId);
    if (!state.boardState[territoryName]) {
        state.boardState[territoryName] = { name: territoryName, spice: 0, forces: {} };
    }
    state.boardState[territoryName].spice = amount;
    await saveState(TestContext.gameId, state);
});



Then('{string} should have {int} Spice', async function (territoryName: string, amount: number) {
    const state = await getState(TestContext.gameId);
    const actual = state.boardState[territoryName]?.spice || 0;
    expect(actual).to.equal(amount);
});

Given('the players are seated as follows:', async function (dataTable: any) {
    if (TestContext.gameId === 0) {
        const initialState: GameState = {
            phase: "Setup",
            turn: 0, stormLocation: 0, factions: [], actionLog: [], auctionQueue: [], currentBid: 0, isBiddingRoundActive: false, boardState: {},
            spiceDeck: [], spiceDiscard: [], treacheryDeck: [], treacheryDiscard: [], nexusActive: false
        };
        const game = await prisma.game.create({
            data: { guildId: "test-storm-3", stateJson: JSON.stringify(initialState), categoryId: "c", actionsChannelId: "a", mapChannelId: "m", tableTalkChannelId: "t" }
        });
        TestContext.gameId = game.id;
    }

    // We mock the seating order in the implementation
    // For now, let's assume standard order A, H, F, E, G, B
    // We populate the factions in that order
    const state = await getState(TestContext.gameId);
    state.factions = [];
    const rows = dataTable.hashes();
    for (const row of rows) {
        state.factions.push({
            faction: row['Faction'] as any,
            playerDiscordId: `d-${row['Faction']}`,
            playerName: row['Faction'],
            spice: 0, reserves: 0, forcesInTanks: 0, leaders: [], traitors: [], hand: []
        });
    }
    await saveState(TestContext.gameId, state);
});

When('the storm moves to sector {int}', async function (targetSector: number) {
    const state = await getState(TestContext.gameId);
    // Force set
    state.stormLocation = targetSector;
    // Trigger First Player Calc
    stormEngine.determineFirstPlayer(state);
    await saveState(TestContext.gameId, state);
    // Storm Engine methods operate on valid object but do not save to DB themselves implicitly unless wired to GameEngine flow.
    // StormSteps uses `stormEngine` directly on `state` object.
    // So saveState IS required here IF stormEngine is pure logic.
    // StormEngine.ts `moveStorm` returns number, updates state object inplace.
    // So for StormSteps, saveState is CORRECT.
});

Then('the First Player should be {string}', async function (factionName: string) {
    const state = await getState(TestContext.gameId);
    const fp = state.factions.find(f => f.playerDiscordId === state.firstPlayerId);
    expect(fp?.faction).to.equal(factionName);
});



Then('the new storm position should be {int}', async function (pos: number) {
    const state = await getState(TestContext.gameId);
    expect(state.stormLocation).to.equal(pos);
});

Then('the "Move Storm" button should be disabled', async function () {
    const state = await getState(TestContext.gameId);
    expect(state.stormMovedThisTurn).to.be.true;
});


